third_party_integration/huggingface_transformers/BootstrapNAS_HF/0001-Modifications-for-BootstrapNAS-usage.patch

```diff
@@ -1775,17 +1791,128 @@ index 170315fe2..9e7df17fc 100644
++        :param matmul_op_metatypes: List of metatypes defining matmul operations.
++        """
++        super().__init__([], linear_op_metatypes)
++        self._n_heads = n_heads
++        self._matmul_op_metatype = matmul_op_metatypes
++
++    def count_flops_and_weights_per_node(self,
++                                         graph: NNCFGraph,
++                                         output_shapes: Dict[NNCFNodeName, int],
++                                         input_channels: Dict[NNCFNodeName, int] = None,
++                                         output_channels: Dict[NNCFNodeName, int] = None,
++                                         kernel_sizes: Dict[NNCFNodeName, Tuple[int, int]] = None,
++                                         op_addresses_to_skip: List[NNCFNodeName] = None) -> \
++        Tuple[Dict[NNCFNodeName, int], Dict[NNCFNodeName, int]]:
++        """
++        Counts the number of weights and FLOPs per node in the model for convolution and fully connected layers.
++
++        :param graph: NNCFGraph.
++        :param output_shapes: Dictionary of output dimension shapes for convolutions and
++            fully connected layers. E.g {node_name: (height, width)}
++        :param input_channels: Dictionary of input channels number in convolutions.
++            If not specified, taken from the graph. {node_name: channels_num}
++        :param output_channels: Dictionary of output channels number in convolutions.
++            If not specified, taken from the graph. {node_name: channels_num}
++        :param kernel_sizes: Dictionary of kernel sizes in convolutions.
++            If not specified, taken from the graph. {node_name: kernel_size}.
++            It's only supposed to be used in NAS in case of Elastic Kernel enabled.
++        :param op_addresses_to_skip: List of operation addresses of layers that should be skipped from calculation.
++            It's only supposed to be used in NAS in case of Elastic Depth enabled.
++        :return Dictionary of FLOPs number {node_name: flops_num}
++                Dictionary of weights number {node_name: weights_num}
++        """
++        input_channels = input_channels or {}
++        output_channels = output_channels or {}
++        kernel_sizes = kernel_sizes or {}
++        op_addresses_to_skip = op_addresses_to_skip or []
++
++        flops, weights = super().count_flops_and_weights_per_node(graph, output_shapes, input_channels, output_channels,
++                                                                  kernel_sizes, op_addresses_to_skip)
++
++        for node in graph.get_nodes_by_metatypes(self._matmul_op_metatype):
++            name = node.node_name
++            if name in op_addresses_to_skip:
++                continue
++
++            # get previous qkv nodes
++            input_nodes = graph.get_previous_nodes(node)
++            while(len(input_nodes) > 0 and input_nodes[0].node_type != 'linear'):
```

Comment:
Since this class is used in HuggingFace transformer, I only consider the matmul operator in self-attention module.